docs/design/2020-06-01-global-kill.md

```diff
@@ -17,28 +17,34 @@ Currently connection ids are local to TiDB instances, which means that a `KILL x
 To support "Global Kill", we need:
 1. Global connection ids, which are unique among all TiDB instances.
 2. Redirect `KILL x` to target TiDB instance, on which the connection `x` is running.
-3. Support both 32 & 64 bits `connId`, to be compatible with legacy 32 bits clients. In this stage, we only design the 64 bits `connId`, and left a `markup` to distinguish between these two kinds.
+3. Support both 32 & 64 bits `connId`. 32 bits `connId` is used on small clusters (number of TiDB instances less than 2048), to be fully compatible with all clients including legacy 32 bits ones, while 64 bits `connId` is used for big clusters. Bit 0 in `connId` is a markup to distinguish between these two kinds.
 
 ## Rationale
 
 #### 1. Structure of `connId`
-##### 64 bits version
+##### 32 bits
+```
+                                      31    21 20               1    0
+                                     +--------+------------------+------+
+                                     |serverId|   local connId   |markup|
+                                     | (11b)  |       (20b)      |  =0  |
+                                     +--------+------------------+------+
+```
+
+##### 64 bits
 ```
   63 62                 41 40                                   1    0    
  +--+---------------------+--------------------------------------+------+
  |  |      serverId       |             local connId             |markup|
  |=0|       (22b)         |                 (40b)                |  =1  |
  +--+---------------------+--------------------------------------+------+
 ```
-##### 32 bits version
-(To be discussed in another RFC)
-```
-                                    31                          1    0
-                                   +-----------------------------+------+
-                                   |             ???             |markup|
-                                   |             ???             |  =0  |
-                                   +-----------------------------+------+
-```
+
+##### Determine 32 or 64 bits
+The key factor is `serverId` (see `serverId` section for detail), which depends on number of TiDB instances in cluster.
+- Choose 32 bits when number of TiDB instances __less than 2048__. Otherwise choose 64 bits.
+- When 32 bits chosen, upgrade to 64 bits when: 1) Fail to acquire `serverId` _(because of occupied)_ continuously __more than 3 times__ _(which will happen when size of cluster is increasing rapidly)_; 2) All `local connId` in 32 bits `connId` are in used (see `local connId` section for detail).
```

Comment:
If we scale out 100 TiDB at the same time (but no more), will it fall into this case?

Reply:
No. But some TiDB will upgrade to 64 bits `connId` because failed to acquire `serverId` from PD (after retried 3 times).